bigquery/google/cloud/bigquery/_pandas_helpers.py

```diff
@@ -187,24 +187,46 @@ def bq_to_arrow_array(series, bq_field):
     return pyarrow.array(series, type=arrow_type)
 
 
-def _columns_and_indexes(dataframe):
+def get_column_or_index(dataframe, name):
+    """Return a column or index as a pandas series."""
+    if name in dataframe.columns:
+        return dataframe[name].reset_index(drop=True)
+
+    if isinstance(dataframe.index, pandas.MultiIndex):
+        if name in dataframe.index.names:
+            return (
+                dataframe.index.get_level_values(name)
+                .to_series()
+                .reset_index(drop=True)
+            )
+    else:
+        if name == dataframe.index.name:
+            return dataframe.index.to_series().reset_index(drop=True)
+
+    raise ValueError("column or index '{}' not found.".format(name))
+
+
+def list_columns_and_indexes(dataframe):
     """Return all index and column names with dtypes.
 
     Returns:
         Sequence[Tuple[dtype, str]]:
             Returns a sorted list of indexes and column names with
-            corresponding dtypes.
+            corresponding dtypes. If an index is missing a name or has the
+            same name as a column, the index is omitted.
     """
+    column_names = frozenset(dataframe.columns)
     columns_and_indexes = []
     if isinstance(dataframe.index, pandas.MultiIndex):
         for name in dataframe.index.names:
-            if name:
+            if name and name not in column_names:
                 values = dataframe.index.get_level_values(name)
                 columns_and_indexes.append((name, values.dtype))
     else:
         if dataframe.index.name:
             columns_and_indexes.append((dataframe.index.name, dataframe.index.dtype))
 
+    # Add columns last so that if you iterate over the list, the column values overwrite any indexes with the same name.
```

Comment:
Is this comment necessary? Don't we omit indexes that have a name equal to one of the column names?

Do we _not_ check if `dataframe.index.name` clashes with a column name on purpose? (we only do it for MultiIndex)